lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+                                            "unexpectedpc_hook",
+                                            M);
+      UnexpectedPCMarker->addFnAttr(Attribute::ReadOnly);
+      UnexpectedPCMarker->addFnAttr(Attribute::NoUnwind);
+      CallInst::Create(UnexpectedPCMarker,
+                       "",
+                       UnexpectedPCCloned->getTerminator());
+    }
+  }
+
+  // Discard it, no longer needed.
+  throwDisposableFunction(OutlinedFunction);
+  return FunctionInfo;
+}
+
+template<class FO>
+FunctionSummary
+CFEPAnalyzer<FO>::milkInfo(ArrayRef<GlobalVariable *> ABIRegisters,
+                           ABIAnalyses::ABIAnalysesResults &ABIResults,
+                           const std::set<GlobalVariable *> &WrittenRegisters,
+                           SortedVector<model::BasicBlock> &CFG,
+                           llvm::Function *F) {
+  using namespace llvm;
+
+  SmallVector<std::pair<CallInst *, uint64_t>, 4> MaybeReturns;
+  SmallSet<std::pair<CallInst *, FunctionEdgeTypeValue>, 4> Result;
+  std::set<GlobalVariable *> CalleeSavedRegs;
+  std::set<GlobalVariable *> ClobberedRegs(ABIRegisters.begin(),
+                                           ABIRegisters.end());
+
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0))) {
+        if (isa<ConstantInt>(CI->getArgOperand(1))) {
+          int64_t FSO = cast<ConstantInt>(CI->getArgOperand(1))->getSExtValue();
+          if (FSO >= 0)
+            MaybeReturns.emplace_back(CI, FSO);
+          else
+            Result.insert({ CI, FunctionEdgeTypeValue::BrokenReturn });
+        } else {
+          // TODO: should this be a longjmp or a brokenreturn?
+          Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      auto *V = dyn_cast<ConstantInt>(CI->getArgOperand(1));
```

Comment:
```suggestion
      auto *StackOffset = dyn_cast<ConstantInt>(CI->getArgOperand(1));
```
Also, document the ratio behind this: we have a "wild" indirect jump but we classify it differently depending on the status of the stack pointer.